src/azul/plugins/repository/tdr/__init__.py

```diff
@@ -314,22 +318,42 @@ def retry_after(self) -> Optional[int]:
         return None
 
 
-class Checksums(NamedTuple):
+@attr.s(auto_attribs=True, kw_only=True, frozen=True)
+class Checksums:
     crc32c: str
-    sha1: str
+    sha1: Optional[str] = None
     sha256: str
-    s3_etag: str
+    s3_etag: Optional[str] = None
 
-    def asdict(self) -> Dict[str, str]:
-        return self._asdict()
+    def to_json(self) -> Dict[str, str]:
+        """
+        >>> Checksums(crc32c='a', sha1='b', sha256='c', s3_etag=None).to_json()
+        {'crc32c': 'a', 'sha1': 'b', 'sha256': 'c'}
+        """
+        return {k: v for k, v in attr.asdict(self).items() if v is not None}
 
     @classmethod
-    def without_values(cls) -> Dict[str, str]:
```

Comment:
In your version of this PR, the tests still injected the optional fields with None values into the JSON representation of the descriptors, even though the real world descriptors will completely lack them. That's a case in point for #2374. It wasn't exactly fun to untangle the "forward" emulation in the test and consolidate that with the "backward" emulation in the CUT without knowing for sure that I didn't break anything.